cpp/src/strings/regex/regcomp.cpp

```diff
@@ -459,16 +459,40 @@ class regex_parser {
     }
 
     // The quantifiers require at least one "real" previous item.
-    // We are throwing an error in these two if-checks for invalid quantifiers.
+    // We are throwing an errors for invalid quantifiers.
     // Another option is to just return CHAR silently here which effectively
     // treats the chr character as a literal instead as a quantifier.
     // This could lead to confusion where sometimes unescaped quantifier characters
     // are treated as regex expressions and sometimes they are not.
     if (_items.empty()) { CUDF_FAIL("invalid regex pattern: nothing to repeat at position 0"); }
 
+    // Check that the previous item can be used with quantifiers.
+    // If the previous item is a capture group, we need to check items inside the
+    // capture group can be used with quantifiers too.
+    // (Note that capture groups can be nested).
+    auto previous_type = _items.back().type;
+    if (previous_type == RBRA) {  // previous item is a capture group
+      // look for matching LBRA
+      auto nested_count = 0;
+      auto lbra_itr =
+        std::find_if(_items.rbegin(), _items.rend(), [nested_count](auto const& item) mutable {
+          nested_count += (item.type == RBRA);
+          return (item.type == LBRA || item.type == LBRA_NC) && (nested_count-- == 0);
+        });
+      // search for the first valid item within the LBRA-RBRA range
+      auto first_valid = std::find_first_of(
+        _items.rbegin() + 1,
+        lbra_itr,
+        valid_preceding_inst_types.begin(),
+        valid_preceding_inst_types.end(),
+        [](auto const item, auto const valid_type) { return item.type == valid_type; });
+      // set previous_type to be checked in next if-statement
+      previous_type = (first_valid != lbra_itr) ? first_valid->type : (--lbra_itr)->type;
```

Comment:
nit: less negation
```suggestion
      previous_type = (first_valid == lbra_itr) ? (--lbra_itr)->type : first_valid->type;
```
It makes sense to me why `first_valid->type` is the right thing to check for here, but why look beyond the left parenthesis? Can this ever dereference past the beginning of the regex?

Reply:
The `lbra_itr` is a reverse iterator so this means `--lbra_itr` is always pointing to a legal item (on it's right) in the vector.

Reply:
Thanks for the clarification!